internal/api/rbac/roles.go

```diff
@@ -0,0 +1,842 @@
+		return
+	}
+	sort.Strings(values)
+	if sa.Annotations == nil {
+		sa.Annotations = map[string]string{}
+	}
+	sa.Annotations[key] = strings.Join(values, ",")
+}
+
+func dropFromClaimAnnotation(sa *corev1.ServiceAccount, key string, values []string) {
+	values = dedupeStringSlice(values)
+	if len(values) == 0 {
+		return
+	}
+	values = removeFromStringSlice(strings.Split(sa.Annotations[key], ","), values)
+	if len(values) == 0 {
+		delete(sa.Annotations, key)
+		return
+	}
+	sort.Strings(values)
+	if sa.Annotations == nil {
+		sa.Annotations = map[string]string{}
+	}
+	sa.Annotations[key] = strings.Join(values, ",")
+}
+
+func buildNewServiceAccount(namespace, name string) *corev1.ServiceAccount {
+	return &corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      name,
+			Annotations: map[string]string{
+				kargoapi.AnnotationKeyManaged: kargoapi.AnnotationValueTrue,
+			},
+		},
+	}
+}
+
+func buildNewRole(namespace, name string) *rbacv1.Role {
+	return &rbacv1.Role{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      name,
+			Annotations: map[string]string{
+				kargoapi.AnnotationKeyManaged: kargoapi.AnnotationValueTrue,
+			},
+		},
+	}
+}
+
+func buildNewRoleBinding(namespace, name string) *rbacv1.RoleBinding {
+	return &rbacv1.RoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      name,
+			Annotations: map[string]string{
+				kargoapi.AnnotationKeyManaged: kargoapi.AnnotationValueTrue,
+			},
+		},
+		Subjects: []rbacv1.Subject{
+			{
+				Kind:      rbacv1.ServiceAccountKind,
+				Namespace: namespace,
+				Name:      name,
+			},
+		},
+		RoleRef: rbacv1.RoleRef{
+			APIGroup: rbacv1.GroupName,
+			Kind:     "Role",
+			Name:     name,
+		},
+	}
+}
+
+func dedupeStringSlice(s []string) []string {
+	seen := make(map[string]struct{}, len(s))
+	deduped := make([]string, 0, len(s))
+	for _, item := range s {
+		if _, ok := seen[item]; !ok {
+			seen[item] = struct{}{}
+			deduped = append(deduped, item)
+		}
+	}
+	return deduped
+}
+
+func removeFromStringSlice(s, items []string) []string {
+	if len(items) == 0 {
+		return s
+	}
+	seen := make(map[string]struct{}, len(items))
+	for _, item := range items {
+		seen[item] = struct{}{}
+	}
+	filtered := make([]string, 0, len(s))
+	for _, item := range s {
+		if _, ok := seen[item]; !ok {
+			filtered = append(filtered, item)
+		}
+	}
+	return filtered
+}
+
+func isKargoManaged(obj metav1.Object) bool {
+	return obj.GetAnnotations()[kargoapi.AnnotationKeyManaged] == kargoapi.LabelTrueValue
+}
+
+func manageableResources(
+	sa corev1.ServiceAccount,
+	roles []rbacv1.Role,
+	rbs []rbacv1.RoleBinding,
+) (*rbacv1.Role, *rbacv1.RoleBinding, error) {
+	gr := schema.GroupResource{
+		Group:    sa.GetObjectKind().GroupVersionKind().Group,
+		Resource: strings.ToLower(sa.GetObjectKind().GroupVersionKind().Kind),
+	}
+	if !isKargoManaged(&sa) {
+		return nil, nil, kubeerr.NewConflict(
```

Comment:
Am wondering if this should be `kubeerr.NewForbidden`? As at present, this results in a slightly confusing error message in certain scenarios:

```console
$ kargo delete role default -p kargo-advanced
Error: already_exists: Operation cannot be fulfilled on serviceaccount "default": ServiceAccount is not Kargo-managed
```

Reply:
Which will translate as a 403, right? I worry people will mis-interpret that as an authz failure.

Reply:
`NewInvalid` then maybe, which yields an [HTTP 422](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/422).

Reply:
I need to double-check, but iirc, connect lacks some status codes we take for granted and invalid is one of them. That said, this may speak to an opportunity to improve the error handling filter. I will look more carefully when I get to my desk.

Reply:
I refreshed my memory on this.

Our error-handling filter maps k8s StatusErrors with code 409 (conflict) to `connect.CodeAlreadyExists` because it's the closest match for a 409 that exists in connect.

That seems to me to be close to correct, because 409 conflict _is_ how I would typically handle something already existing. In fact, kuberr.NewAlreadyExists wraps a 409.

So the long and short of it is while I _think_ the circumstances here do meet the definition of a conflict, we also know the technical constraints we're dealing with will obscure that error and make it look like something it is not.

I agree with your suggestion that 422 (invalid) would be a good alternative, but `kubeerr.NewInvalid()` looks fairly onerous, requiring you to pinpoint exactly which _fields_ have issues, and I'm not sure the circumstances here permit us to do that.

I think another acceptable option could be 400 (bad request), which is also arguably true, when a user has requested an update or delete of something Kargo cannot manage. (Especially given that Kargo Roles now have a field indicating whether they are Kargo-managed or not, a client has the ability to know ahead of time that they ought not make a request to mutate such a Kargo Role.) I think this is the direction I am inclined to take... unless you talk me out of it. 😄

Reply:
We talked offline and agreed on 400.